Mathlib/FieldTheory/Perfect.lean

```diff
@@ -180,3 +180,34 @@ instance toPerfectRing (p : ℕ) [hp : Fact p.Prime] [CharP K p] : PerfectRing K
   exact minpoly.degree_pos ha
 
 end PerfectField
+
+namespace Polynomial
+
+variable {R : Type*} [CommRing R] [IsDomain R] {p n : ℕ} [Fact p.Prime] [CharP R p]
+  [PerfectRing R p] {f : R[X]}
+
+open Multiset
+
+theorem roots_expand : (expand R p f).roots = p • f.roots.map (frobeniusEquiv R p).symm := by
+  classical
+  refine ext' fun r ↦ ?_
+  rw [count_roots, rootMultiplicity_expand, ← count_roots, count_nsmul, count_map,
+      count_eq_card_filter_eq]; congr; ext
+  exact (frobeniusEquiv R p).eq_symm_apply.symm
+
+theorem roots_expand_pow :
+    (expand R (p ^ n) f).roots = p ^ n • f.roots.map (frobeniusEquiv R p).symm ^[n] := by
+  induction' n with n ih generalizing f; · simp
+  rw [pow_succ', expand_mul, ih, mul_smul, roots_expand,
+      Multiset.map_nsmul, Multiset.map_map, iterate_succ]
+
+theorem roots_expand_map_frobenius : (expand R p f).roots.map (frobenius R p) = p • f.roots := by
```

Comment:
Is it possible to add a version without requiring `PerfectRing R p`, which states that `(expand R p f).roots.map (frobenius R p) ≤ p • f.roots`? (Similarly, there should be a `rootsExpandToRoots` without requiring `PerfectRing R p` in #9271, but I forgot to added. It should state that `(expand R p f).roots.toFinset -> f.roots.toFinset`.)

Reply:
Looks like that in order to state `roots_expand` without requiring perfect ring, `f.roots.map (frobeniusEquiv R p).symm` should be replaced by `f.roots.preimage' (frobenius R p) sorry` where `Multiset.preimage'` is the preimage of a multiset under the condition that its preimage as a set is finite (`Set.Finite`). Similar to `Finset.preimage` but with condition `Set.InjOn` replaced by `Set.Finite`.

Reply:
It's probably not worth introducing a new def for this; it won't be computable, it's not clear it's useful elsewhere, and you'll need to develop APIs around it. Also I don't see why you wouldn't directly use `rootMultiplicity_expand`; it's the main theorem and the others are consequences provided for convenience, because they have compact statements. If the statement is too complicated it's probably not worth it. However I agree adding `(frobenius R p) ≤ p • f.roots` is a good idea and we should be satisfied with it for the non-perfect case.